### DISCUSSION TITLE
Scanlist states

### DISCUSSION THREAD

**Discussion by oisact** [AubsUK/uv-k5-firmware-custom#9] [Q&A]
It isn't clear what the differences are between the three scanlist states. They seem to be:

- Enabled Active
- Enabled Inactive
- Disabled

It's not clear what the difference is between Enabled Inactive and Disabled, because neither the Enabled Inactive channels nor Disabled channels are scanned.

**Comment by AubsUK** [comment]
I'm not sure I follow.  Does this make sense:
![image](https://github.com/AubsUK/uv-k5-firmware-custom/assets/68870168/6c6582e7-c1d3-44f8-b8b9-9bd6537f2bf3)

    * If a ScanList is disabled, it will show as an underscore (_) 
    * If a ScanList is enabled, it will show as an:
        * Asterisk (*) if it is not the active ScanList
        * Number (0-9) if it is the active ScanList

**Reply by oisact** [reply]
Say I have two scan lists, 1 and 2.  I start 1 first, then enable 2.  Thus I have `-1*-------` That never changes while scanning. If I press `1` and disable that scan list, then I see `--2-------`  Now if I press `1` I see `_*2-------` and that never changes.

Is the expected behavior that it will scan list 1 showing `-1*`, and when that is done scanning the memories in 1 then it would show `-*2` and scan the memories in list 2, and continuously alternate back and forth? Because for me it only scans one list, the one that is a number, and there is no difference between the `*` and `-` lists as neither are scanned.

Is there some setting that disables scanning multiple lists?

**Reply by AubsUK** [reply]
![Scanning4and5](https://github.com/AubsUK/uv-k5-firmware-custom/assets/68870168/68a3d2c9-3029-4c7e-8e1f-64576a19f9c5)
Hopefully that shows what it should do.  I have 8 channels in ScanList 4, so it scans it really quickly.  I have 32 channels in ScanList 5, so it takes a bit longer.

Do you have any memories stored in both ScanLists?
If you have the latest version (00.03), go to Menu 17 (you can press `M` then `1` then `7` and it'll jump straight to menu 17) Press `M` to get into the lists, then press the number (`0`..`9`) of the ScanList you want to check. This will list all channels stored in that ScanList, so you can scroll through to confirm they have been added.
If you have an older version, it'll be Menu 20-29, one menu item for each ScanList.

**Reply by AubsUK** [reply]
(and sorry I didn't reply 2 days ago. I didn't get any notifications which is disappointing, or I'd have been straight on it!)

**Comment by oisact** [comment]
I just flashed to V3, and re-uploaded from Chirp. I see the same behavior as before.   I have 8 memories in Scan List 1 (1-3, 10-14), and 14 in Scan List 5 (100-106, 108-114). It never switches from one Scan List to another. Whichever Scan List I activate first is the only list scanned. Whether I have two active when I begin scanning, or have one active and bring another active after scanning has started.

Here is a video. Hopefully it is good enough you can see the behavior. In this I let it run for a bit, then I toggle back and forth between the two Scan Lists  manually by deactivating the one being scanned.
https://www.youtube.com/watch?v=mfKy4na0i7k

**Reply by AubsUK** [reply]
I think I might now understand what's up.  To have multiple ScanLists and allow all channels to be in more than one ScanList, I've utilised some of the storage used for the NAME (it was using 16 bytes to store 10 bytes of data, so I've pinched 2 bytes).  Did you follow the note at the top of the ReadMe:

> [!IMPORTANT]
> For first use, I would highly recommend taking (and saving) a backup of your existing radio configuration through the official [CHIRP release](https://chirp.danplanet.com/projects/chirp/wiki/Download), using whatever firmware currently flashed.
> 1. Once you have that, flash this firmware - See [Releases](https://github.com/AubsUK/uv-k5-firmware-custom/releases)
> 2. Perform a factory reset (hold `PTT` + `SIDE BUTTON 1` while turning on the radio >> Menu 77 `Reset` >> `ALL`) to clear some EEPROM locations that other firmware configures slightly differently
> 3. Use CHIRP with [Aubs UV-K5 CHIRP Driver](https://github.com/AubsUK/uvk5-chirp-driver) to read from radio
> 4. Copy/paste the channels from the backup in CHIRP to the downloaded configuration
> 5. Turn ScanLists off/on as required
> 6. Add channels to ScanLists as required
> 7. Finally upload back to the radio and you should see the model as `UV-K5 Aubs`

**Reply by oisact** [reply]
Yes. After flashing I reset the radio. Then I downloaded the empty state from the radio into Chirp with the custom driver loaded. Then I copied my memories from another file into that new one.
I'm not certain of steps 5 and 6 as relating to the radio and / or chirp? Does that mean upload to the radio, turn on scanlists, then in chirp assign channels to scanlists and download to radio?

**Reply by AubsUK** [reply]
Steps 5/6 don't really matter.  It was intended to set them up in CHIRP before uploading to radio, but it doesn't really matter if you do it there, or do it when scanning.

To help us diagnose, if you're comfortable to...

1. In CHIRP, go to Help >> Developer Mode
2. Restart CHIRP
3. In CHIRP, go to Radio >> Download from radio
4. Select the Browser tab
(wait, it'll take a while)
6. On the left, click on ChannelScanLists.  It'll expand all the channel numbers starting at 0 (0 = Channel 1, 1 = Channel 2, etc.)
7. On the left, select a channel number that you know will be empty (if you know Channel 34 is empty, select 33)
8. You will see on the right a list of all the memory locations starting at ScanList7 to ScanList0, then lower down ScanList9 and Scanlist8.

Are they all 1s or 0s ?

Here's an example for my Channel20 (memory location 19) which is empty:

![image](https://github.com/AubsUK/uv-k5-firmware-custom/assets/68870168/c5de4a01-34f3-45e2-9846-9a5e909cd3b6)


Unless you want to keep Developer Mode on, it would be best to turn it off again (will need a restart of CHIRP).

**Reply by oisact** [reply]
They are all `1`s.

**Reply by AubsUK** [reply]
Can you try and factory reset it again, because [this](https://github.com/AubsUK/uv-k5-firmware-custom/blob/main/settings.c#L482) bit of code should set the memory locations to `0`'s:
```
		// If we're in the second half of the 'Memory Name' location, set all bits to 0
		if (i >= 0x0F50 && i < 0x1C00) { EEPROM_WriteBuffer(i, TemplateB);  }
		// Otherwise, set all bits to 1
		else { EEPROM_WriteBuffer(i, TemplateA); }
```
Which means all the ScanLists for each channel should be off.

If you're interested, the code block is not quite as it should be, because it currently sets all 8 bytes zeros, but the first two bytes are part of the Channel Name and should be 1s.  The next version will have this code to do it properly:

```
		// Write 8 bytes to 1s
		EEPROM_WriteBuffer(i, TemplateOnes);
		// If we're in the 'Memory Name' location, increment the loop counter by 8 and set the second half bytes to 1s for first two and 0 for the remaining (for the ScanList locations)
		if (i >= 0x0F50 && i < 0x1C00) {
			i += 8;
			EEPROM_WriteBuffer(i, TemplateOneZeros);
		}
```



### The next version (00.04 when released) will also have an additional check that the channel has been programmed when scanning, so it'll ignore the 1s that you have and most likely mean a factory reset isn't needed after flashing from other firmware as long as you set up the channels to be in the right ScanLists either in CHIRP or the radio.

**Comment by oisact** [comment]
So the fundamental problem here may be that I reset the radio wrong. I powered it on while holding PTT and the side button, and got the "RELEASE ALL KEYS" message. However, I did not go into the menu and reset there, as I thought that was just another way to achieve the same reset.  You may want to tweak your wording and say "then enter settings and go to menu...". It was my misinterpretation but I thought the power on method and menu method achieved the same thing.

I did the full reset, but now I have to create another CHIRP file because that is the primary issue, and not anything to actually do with the radio.

**Reply by AubsUK** [reply]
Thank you, I'll update the notes.

So if you've already got a CHRP file with all your programmed frequencies set up:

1. Open CHIRP with my driver
2. Select File >> Open and find your existing list (so it shows as a tab at the top of CHIRP)
3. Radio >> Download from radio (which will download a blank list from the radio with all the configuration for my version)
4. Switch to your existing list
5. Click on one channel
6. Press Ctrl+A to select all
7. Press Ctrl+C to copy all
8. Switch to the blank list and click on the first one (it'll have a frequency already entered)
9. Press Ctrl+V to paste
10. Change the ScanLists to put the channels in one or more ScanList (I find it easier to click in the first one, enter the ScanList number(s), press enter which takes you to the next one down, type in the next channel's ScanList number(s), press enter, and keep doing that.  Takes seconds to do them all.)
11. Radio >> Upload to radio

**Comment by oisact** [comment]
Okay, that did it. I'm seeing the correct Scan List behavior now. Thanks for the assistance!

**Reply by AubsUK** [reply]
No, thank you for your perseverance, and willingness to try the firmware changes.  I'm pleased you got it sorted.  I've also updated the "_Important_" instructions to make the factory reset clearer.

I'm really happy to take feedback and make positive changes to the firmware.
